Deriving fluids from liquids

4Denthusiast opened the pull request:
Rather than declaring a water prefab in Fluid, derive a fluid for each block with the liquid attribute. This replaces the fluid renderer with just a BufferedImage from the block's tile, and adds more information (fluid-block correspondences and display names). Components can still be attached to the fluid using the block's prefab, and fluids which can't be placed in the world (e.g. IRLCorp:steam) can still be declared separately.

This PR is based on the same branch as the PR for removing liquid from the world when it's picked up with a fluid container item, because there's some overlap and both are pre-requisites for another planned extension to allow placing liquids in the world from containers.

Modules which depend on Fluid are likely to need corresponding updates, particularly because of the changes in fluid rendering, and the replacement of Fluid:water with CoreAssets:Water. I have some ready to go that I'll upload shortly, but I still need to check if there are any others that I don't already have the sources for. There's also a small engine change necessary, just adding BlockTile to the API.

Pretty much the only direct effect of this is that you can now pick up lava and FlowingLiquids:DebugLiquid in containers, and all of these liquids are textured rather than being single colours, but it tidies things up and enables further features later.

4Denthusiast committed:
Remove liquid from the world when filling container items.

4Denthusiast committed:
Respond to review comments.

4Denthusiast committed:
Make FlowingLiquids dependency optional.

When picking up liquid with FL not enabled, the block is just entirely
removed or left in place, randomly with probabilities corresponding to
the amount of liquid removed.

4Denthusiast committed:
Derive fluids from liquid blocks

4Denthusiast commented:
That should be all of the dependency updates done.

pollend commented on src/main/java/org/terasology/fluid/system/FluidRegistry.java:
```diff
@@ -23,15 +29,33 @@
      * @param fluidType The fluid type
      */
-    FluidRenderer getFluidRenderer(String fluidType);
+    BufferedImage getFluidTexture(String fluidType);
+
+    String getDisplayName(String fluidType);
+
+    Block getCorrespondingLiquid(String fluidType);
+
+    String getCorrespondingFluid(Block liquid);
 
+    Prefab getPrefab(String fluidType);
 }
```
do we need to add javadocs to these methods?

4Denthusiast replied:
Most of the javadocs existed in the implementation already, so I copied them over to the interface (and added one for getPrefab).

4Denthusiast committed:
Add more javadoc.

4Denthusiast committed:
Merge branch 'develop' into derivingFluids

Cervator commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -52,6 +57,34 @@
+    
+    @In
+    private ExtraBlockDataManager extraDataManager;
+    private int flowIx;
```
Is there any specific meaning to "Ix" here? Might be some convention I'm unfamiliar with

skaldarnar replied:
I asked the same in the other PR 🙈 : It stands for `index`. As @Cervator now is the second one having trouble with this abbreviation I'd like to ask you to change this to `flowIndex` for more clarity (the additional characters won't hurt us 🙃 ):
```suggestion
    private int flowIndex;
```

4Denthusiast replied:
Okay, will do. How do you make suggested changes like that on Github?

Cervator replied:
Just a cool new feature during a review - there's a button for "Insert Suggestion" 👍 

![image](https://user-images.githubusercontent.com/1063833/88691303-50706f00-d0c2-11ea-8fdf-8cf146b8c855.png)

Cervator commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -52,6 +57,34 @@
+     * If one block is 1m across, the fluid units are litres. This works reasonably
+     * sensibly with the pre-existing container sizes in ManualLabor.
+     */
+    private static final float FLUID_PER_BLOCK = 1000;
```
Since this is a constant should it be up by the top of the class instead?

skaldarnar replied:
How can we make the Checkstyle results more visible on PR builds 🤔 🤓

4Denthusiast replied:
Checkstyle doesn't even complain about this. I've moved it anyway.

4Denthusiast replied:
Actually I think Checkstyle might just not be running for some reason. It's also not complaining about some other stuff that's definitely against the rules.

Cervator replied:
I think I've gotten used to the IDE warnings mixing together Checkstyle and some IntelliJ recommendations, not even sure which one covers this. With the latest Jenkins rebuild we do get this nice check:

![image](https://user-images.githubusercontent.com/1063833/88690484-54e85800-d0c1-11ea-8771-3351240e4842.png)

Although it is still pretty mild and hidden, I think we can change the severity config, add gates, etc to make it look more alarming or block PRs. Another option is to have a bot make outright comments or something (that just might get spammy). Finally I *do* believe there's another level to the fancy Cloudbees Jenkins GitHub PR thing that can be enabled with a full GitHub app - that _may_ be outright line coverage richly outlined in the PR which would be amazing, but I'm not sure as I haven't had time yet to dig deeper.

Following the link that gets posted you get all the Checkstyle details, can even drill down to where this exact one is covered :-)

http://jenkins.terasology.io/teraorg/job/Terasology/job/Modules/job/F/job/Fluid/job/PR-19/3/checkstyle/packageName.-1493123546/fileName.1389089918/type.1614765908/

![image](https://user-images.githubusercontent.com/1063833/88690922-ce804600-d0c1-11ea-8d54-093268f05b2a.png)

So we might be in business, just needing to make it all more visible

Cervator commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -95,21 +129,61 @@
+            blockAmount *= LiquidData.MAX_HEIGHT;
+        }
+        int liquidLevel = randomRound(blockAmount);
+        if(liquidLevel == 0) {
```
Trivial: Missing space (checkstyle IDE plugin ftw!)

Cervator commented on src/main/java/org/terasology/fluid/system/FluidClientSystem.java:
```diff
@@ -47,6 +47,8 @@
 public class FluidClientSystem extends BaseComponentSystem {
 
+    @In
+    private FluidRegistry fluidRegistry;
```
Trivial: No empty line after this, but yep empty line after on a similar change

Cervator reviewed (approved):
Left some trivial code style comments, overall looks good to me, I especially like the level of javadoc/comments 👍 

Have not tested, but poked for review elsewhere. 

Curious if this interacts with #20? Both PRs started around the same time

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -95,21 +129,61 @@
         }
     }
+    
+    private float getLiquidInBlock(Vector3i pos) {
```
Is this supposed to get the actual liquid/fluid in the block at position `pos`? If so, please add JavaDoc stating that this method assumes that the block at the given position actually has/is a fluid block (at least in the non-FlowingLiquids case).

Also, I'm confused why we mix up **fluid** and **liquid** here 🙈

4Denthusiast replied:
I've been distinguishing between liquids, blocks with the `liquid` flag, and fluids, things registered with the `FluidRegistry` that can appear in containers and things. As this is converting between fluid and liquid it's a bit of an edge case.

skaldarnar replied:
I see. Sooo much potential for JavaDoc and module documentation in general ^^

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -95,21 +129,61 @@
+        }
+    }
+    
+    // Assumes that the block is already a liquid, so the block ID doesn't need to be set unless the liquid is entirely removed.
```
Please make this regular JavaDoc syntax and also describe the restrictions/assumptions on the `fluidAmount` parameter

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidCommonSystem.java:
```diff
@@ -36,14 +47,24 @@
+        for (ResourceUrn blockUrn : Assets.list(BlockFamilyDefinition.class)) {
+            Optional<BlockFamilyDefinition> maybeDefinition = Assets.get(blockUrn, BlockFamilyDefinition.class);
+            if (maybeDefinition.isPresent()) {
+                SectionDefinitionData blockData = maybeDefinition.get().getData().getBaseSection();
```
this code can be more idiomitic in the usage of `Optional`

```suggestion
            maybeDefinition.ifPresent(definition -> {
                SectionDefinitionData blockData = definition.getData().getBaseSection();
```

See [12 recipes for using the Optional class as it’s meant to be used](https://blogs.oracle.com/javamagazine/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used) and [The Java Optional class: 11 more recipes for preventing null pointer exceptions](https://blogs.oracle.com/javamagazine/the-java-optional-class-11-more-recipes-for-preventing-null-pointer-exceptions) for more information on this topic 😉

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidRegistryImpl.java:
```diff
@@ -28,28 +36,102 @@
 @Share(FluidRegistry.class)
 public class FluidRegistryImpl extends BaseComponentSystem implements FluidRegistry {
-    private Map<String, FluidRenderer> fluidRenderers = new HashMap<>();
+    private Map<String, BufferedImage> fluidTextures = new HashMap<>();
+    private Map<String, String> displayNames = new HashMap<>();
+    private Map<String, Block> fluidLiquidMap = new HashMap<>();
+    private Map<Block, String> liquidFluidMap = new HashMap<>();
```
Consider using `Name` class instead of String in these maps to use our standard for case-insensitive strings. I think you've put the `.toLowerCase()` everywhere, but it's easy to miss those and run into nasty case-debugging 🙈

4Denthusiast replied:
This turned out to require loads of changes in other modules too, as fluid names were already strings even before this PR, but I agree that it makes sense. It just occurred to me after making all the changes that I could maybe have just changed FluidRegistryImpl internally, but I might as well go ahead with it now.

4Denthusiast replied:
There was a problem with replacing `String` with `Name` for the fluid IDs everywhere, that in some cases that doesn't work with serialization, so I ened up just changing this internally in `FluidRegistryImpl`.

skaldarnar replied:
Sounds fair - I'm also not sure whether `Name` is the best solution for cases like this, but I'd say it's a start. Starting off internally is totally fine, thanks for making this change! 💚 

`Name` _should_ have proper type handlers, thus I'm a bit surprised about serialization issues 🤔

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidRegistryImpl.java:
```diff
@@ -28,28 +36,102 @@
+        if (liquidBlock != null) {
+            return liquidBlock.getPrefab().orElse(null);
+        } else {
+            return CoreRegistry.get(PrefabManager.class).getPrefab(fluidType);
```
Why do we need to get the `PrefabManager` from the `CoreRegistry` here? Why not use `@In PrefabManager prefabManager;`?

4Denthusiast replied:
The `PrefabManager` may be initialised after the `FluidRegistry`, in which case `@In` wouldn't work, but it should be initialised before this method is actually called. This came up during testing.

skaldarnar reviewed (changes requested):
Overall the change look fine to me, but I'd like to request some changes nevertheless, mostly code or documentation improvements 🙃

4Denthusiast commented:
> Curious if this interacts with #20? Both PRs started around the same time

There was a little bit of a merge conflict, but #20 changes the item icons using a different system so actually not that much. I resolved the conflict and tested them together when #20 was merged.

4Denthusiast committed:
Respond to review. Fix replacing fluids in container items.

4Denthusiast commented:
I think I've fixed everything from the reviews so far. Also I found and fixed a bug where trying to pick up a different type of liquid than the one that the container item was already partially full with would overwrite it or something. This previously wouldn't have occurred because until now there was only one liquid that could be picked up.
